lib/common/fs-operation.js

```diff
@@ -0,0 +1,251 @@
+// Copyright 2016, EMC, Inc.
+
+'use strict';
+
+module.exports = FsOperationFactory;
+
+FsOperationFactory.$provide = 'fs-operation';
+FsOperationFactory.$inject = [
+    'assert',
+    'Logger',
+    'fs',
+    'path',
+    'mkdirp',
+    'child_process',
+    'Promise',
+    '_'
+];
+
+/**
+ * FsOperationFactory returns a FsOperation instance.
+ * @private
+ */
+function FsOperationFactory(
+    assert,
+    Logger,
+    fs,
+    path,
+    mkdirp,
+    childProcess,
+    Promise,
+    _
+) {
+    var logger = Logger.initialize("fs-op");
+
+    function FsOperation() {
+    }
+
+    FsOperation.prototype.getFileNameFromStr = function getFileNameFromStr(str) {
+        var splited = str.split('/');
+        if(splited.length > 0){
+            return splited[splited.length - 1];
+        } else {
+            return '';
+        }
+    };
+
+    FsOperation.prototype.getIsoNameFromStr = function getIsoNameFromStr(str) {
+        var name = this.getFileNameFromStr(str);
+        if (name.endsWith('.iso')) {
+            return name;
+        } else {
+            throw new Error('ISO path should end with .iso');
+        }
+    };
+
+    FsOperation.prototype.getIsoNameFromWebLink =
+        function getIsoNameFromWebLink(osName, osVersion, link) {
+            var iso = this.getIsoNameFromStr(link);
+            return iso? iso: osName + '-' + osVersion;
+        };
+
+    FsOperation.prototype.mkdirSync = function mkdirSync(path) {
+        try {
+            // fs.mkdirSync(path);
+            logger.info('Creating folder: ' + path);
+            mkdirp.sync(path);
+        } catch (err) {
+            logger.error('Error creating folder ' + path + ' Error code: ' + err);
+            throw new Error('Error creating folder ' + path + ' Error code: ' + err);
+        }
+    };
+
+    FsOperation.prototype.checkPathWritable = function checkPathWritable(path) {
+        try {
+            fs.accessSync(path, fs.R_OK | fs.W_OK);
+            return true;
+        } catch (err) {
+            logger.error('Path is not writable: ' + path + ' Error code: ' + err);
+            throw new Error('Path is not writable: ' + path + ' Error code: ' + err);
+        }
+    };
+
+    FsOperation.prototype.checkPathReadable = function checkPathReadable(path) {
+        try {
+            fs.accessSync(path, fs.R_OK);
```

Comment:
I understand the `sync` version API will make the code more simple, but it will also impact the concurrency performance especially in scale environment. "Async" is the marrow of node.js